CMakeLists.txt

```diff
@@ -0,0 +1,374 @@
+				OUTPUT_STRIP_TRAILING_WHITESPACE)
+	endif()
+else()
+	execute_process(COMMAND cat .version
+			OUTPUT_VARIABLE SRCVERSION
+			WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
+			OUTPUT_STRIP_TRAILING_WHITESPACE)
+endif()
+add_definitions(-DSRCVERSION="${SRCVERSION}")
+add_definitions(-DPMEMFILE_MAJOR_VERSION=${VERSION_MAJOR})
+add_definitions(-DPMEMFILE_MINOR_VERSION=${VERSION_MINOR})
+if(NOT WIN32)
+	add_definitions(-DUSE_VALGRIND)
+endif()
+
+include(CheckCCompilerFlag)
+include(CheckCXXCompilerFlag)
+include(GNUInstallDirs)
+find_package(PkgConfig QUIET)
+
+include_directories(include)
+
+set(CMAKE_C_STANDARD 99)
+set(CMAKE_CXX_STANDARD 11)
+
+macro(add_c_flag flag)
+	if(${ARGC} GREATER 2)
+		set(check_name ${ARGV2})
+	else()
+		string(REPLACE - _ flag2 ${flag})
+		set(check_name C_HAS_${flag2})
+	endif()
+
+	check_c_compiler_flag("${flag}" ${check_name})
+
+	if (${${check_name}})
+		if (${ARGC} EQUAL 1)
+			set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${flag}")
+		else()
+			set(CMAKE_C_FLAGS_${ARGV1} "${CMAKE_C_FLAGS_${ARGV1}} ${flag}")
+		endif()
+	endif()
+endmacro()
+
+macro(add_cxx_flag flag)
+	if(${ARGC} GREATER 2)
+		set(check_name ${ARGV2})
+	else()
+		string(REPLACE - _ flag2 ${flag})
+		set(check_name CXX_HAS_${flag2})
+	endif()
+
+	check_cxx_compiler_flag(${flag} ${check_name})
+
+	if (${${check_name}})
+		if (${ARGC} EQUAL 1)
+			set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${flag}")
+		else()
+			set(CMAKE_CXX_FLAGS_${ARGV1} "${CMAKE_CXX_FLAGS_${ARGV1}} ${flag}")
+		endif()
+	endif()
+endmacro()
+
+add_c_flag(-Wall)
+add_c_flag(-Wmissing-prototypes)
+add_c_flag(-Wpointer-arith)
+add_c_flag(-Wunused-macros)
+add_c_flag(-Wsign-conversion)
+add_c_flag(-Wsign-compare)
+add_c_flag(-Wconversion)
+add_c_flag(-Wunreachable-code-return)
+add_c_flag(-Wmissing-variable-declarations)
+
+add_cxx_flag(-Wall)
+add_cxx_flag(-Wpointer-arith)
+add_cxx_flag(-Wunused-macros)
+add_cxx_flag(-Wsign-conversion)
+add_cxx_flag(-Wsign-compare)
+add_cxx_flag(-Wconversion)
+add_cxx_flag(-Wunreachable-code-return)
+add_cxx_flag(-Wmissing-variable-declarations)
+
+add_c_flag(-pthread)
+add_c_flag(-fno-common)
+add_cxx_flag(-pthread)
+add_cxx_flag(-fno-common)
+
+check_c_compiler_flag(-Wl,-z,relro LINKER_HAS_RELRO)
+if(LINKER_HAS_RELRO)
+	set(CMAKE_LD_FLAGS  "${CMAKE_LD_FLAGS}  -Wl,-z,relro")
+endif()
+
+check_c_compiler_flag(-Wl,--warn-common LINKER_HAS_WARN_COMMON)
+if(LINKER_HAS_WARN_COMMON)
+	set(CMAKE_LD_FLAGS  "${CMAKE_LD_FLAGS} -Wl,--warn-common")
+endif()
+
+add_c_flag(-O0 DEBUG)
+add_c_flag(-ggdb DEBUG)
+add_c_flag(-DDEBUG DEBUG)
+
+add_cxx_flag(-O0 DEBUG)
+add_cxx_flag(-ggdb DEBUG)
+add_cxx_flag(-DDEBUG DEBUG)
+
+add_c_flag("-U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2" RELEASE C_CAN_FORTIFY)
+add_cxx_flag("-U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2" RELEASE CXX_CAN_FORTIFY)
+
+add_c_flag(-D_LARGEFILE64_SOURCE)
+add_cxx_flag(-D_LARGEFILE64_SOURCE)
+
+option(SANITIZERS "enable AddressSanitizer and UndefinedBehaviorSanitizer (debugging)" OFF)
+
+if(SANITIZERS)
+	if(BUILD_LIBPMEMFILE)
```

Comment:
IIRC sanitizers override some libc functions and invoke syscalls without going through libc.